### PR TITLE
k8s-infra-prow-build: Add monitoring

### PR DESCRIPTION
**Pull request by koksay**
Adds monitoring components for the `k8s-infra-prow-build` cluster. 
Mainly copied from `eks-prow-build-cluster` [configuration](https://github.com/kubernetes/k8s.io/tree/main/infra/aws/terraform/prow-build-cluster/resources/monitoring)

Please note that, the `alertmanager-slack-token.yam` file has invalid data for this cluster. It will be updated later when the external secrets are in place.

### CODE REVIEW

File: infra/gcp/terraform/k8s-infra-prow-build/prow-build/resources/monitoring/alertmanager-main/alertmanager-config.yaml [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+    # Default group criteria. This might need some refactor if we add more
+    # alerting rules.
+    groupBy: ['alertname', 'node']
+  receivers:
+  - name: 'slack'
+    slackConfigs:
+    - apiURL:
+        name: alertmanager-k8c-slack-token
+        key: url
+      channel: '#alerting-cncf-prod'
```
Comment (ameukam) [inline infra/gcp/terraform/k8s-infra-prow-build/prow-build/resources/monitoring/alertmanager-main/alertmanager-config.yaml:49]: Probably drop the entire alerting config for the moment. we want metrics visualization first.

File: infra/gcp/terraform/k8s-infra-prow-build/prow-build/resources/monitoring/grafana/dashboards/kube-state-metrics.yaml
Comment (xmudrii) [inline infra/gcp/terraform/k8s-infra-prow-build/prow-build/resources/monitoring/grafana/dashboards/kube-state-metrics.yaml:1]: I'd go only with `prow.yaml` for this cluster.

File: infra/gcp/terraform/k8s-infra-prow-build/prow-build/resources/monitoring/grafana/dashboards/node-exporter-full.yaml
Comment (xmudrii) [inline infra/gcp/terraform/k8s-infra-prow-build/prow-build/resources/monitoring/grafana/dashboards/node-exporter-full.yaml:1]: I'd go only with `prow.yaml` for this cluster.

File: infra/gcp/terraform/k8s-infra-prow-build/prow-build/resources/monitoring/grafana/datasources.yaml [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+    - access: proxy
+      isDefault: true
+      name: Prometheus Main
+      type: prometheus
+      url: http://prometheus-operated.monitoring:9090
+    - access: proxy
+      isDefault: false
+      name: Loki
+      type: loki
+      url: http://loki-stack.monitoring:3100
```
Comment (xmudrii) [inline infra/gcp/terraform/k8s-infra-prow-build/prow-build/resources/monitoring/grafana/datasources.yaml:19]: I don't see that we have Loki running in this cluster, please correct me if I'm wrong.

File: infra/gcp/terraform/k8s-infra-prow-build/prow-build/resources/monitoring/grafana/deployment.yaml [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+        - name: test-infra
+          configMap:
+            name: test-infra-dashboard
+      nodeSelector:
+        node-group: stable
+      tolerations:
+      - effect: NoSchedule
+        key: node-group
+        operator: Equal
+        value: stable
```
Comment (xmudrii) [inline infra/gcp/terraform/k8s-infra-prow-build/prow-build/resources/monitoring/grafana/deployment.yaml:128]: We don't have stable nodes in the GKE cluster.

File: infra/gcp/terraform/k8s-infra-prow-build/prow-build/resources/monitoring/grafana/lb.yaml
Comment (xmudrii) [inline infra/gcp/terraform/k8s-infra-prow-build/prow-build/resources/monitoring/grafana/lb.yaml:1]: Let's remove Service for now, we'll expose it in a follow up PR.

File: infra/gcp/terraform/k8s-infra-prow-build/prow-build/resources/monitoring/grafana/secret.yaml
Comment (xmudrii) [inline infra/gcp/terraform/k8s-infra-prow-build/prow-build/resources/monitoring/grafana/secret.yaml:1]: Is this empty secret/file expected?

File: infra/gcp/terraform/k8s-infra-prow-build/prow-build/resources/monitoring/grafana/service.yaml
Comment (xmudrii) [inline infra/gcp/terraform/k8s-infra-prow-build/prow-build/resources/monitoring/grafana/service.yaml:1]: Any idea why do we need this Service (I know it's probably just pasted from the EKS cluster)?

File: infra/gcp/terraform/k8s-infra-prow-build/prow-build/resources/monitoring/kube-state-metrics/deployment.yaml [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+                - ALL
+            readOnlyRootFilesystem: true
+            runAsUser: 65534
+      nodeSelector:
+        node-group: stable
+      tolerations:
+      - effect: NoSchedule
+        key: node-group
+        operator: Equal
+        value: stable
```
Comment (xmudrii) [inline infra/gcp/terraform/k8s-infra-prow-build/prow-build/resources/monitoring/kube-state-metrics/deployment.yaml:58]: We don't have stable nodes in the GKE cluster.

File: infra/gcp/terraform/k8s-infra-prow-build/prow-build/resources/monitoring/prometheus-main/prometheus.yaml [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+        resources:
+          requests:
+            storage: 300Gi
+  nodeSelector:
+    node-group: stable
+  tolerations:
+  - effect: NoSchedule
+    key: node-group
+    operator: Equal
+    value: stable
```
Comment (xmudrii) [inline infra/gcp/terraform/k8s-infra-prow-build/prow-build/resources/monitoring/prometheus-main/prometheus.yaml:54]: We don't have stable nodes in the GKE cluster.

File: infra/gcp/terraform/k8s-infra-prow-build/prow-build/resources/monitoring/prometheus-main/prometheus.yaml [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+  logLevel: info
+  logFormat: logfmt
+  retention: 7d
+  scrapeInterval: 15s
+  securityContext:
+    fsGroup: 0
+    runAsNonRoot: false
+    runAsUser: 0
+  storage:
+    volumeClaimTemplate:
```
Comment (xmudrii) [inline infra/gcp/terraform/k8s-infra-prow-build/prow-build/resources/monitoring/prometheus-main/prometheus.yaml:43]: We should check if the default StorageClass on GKE is "good-enough" for Prometheus (e.g. not too expensive, solid performance, etc.).

**Review by xmudrii — CHANGES_REQUESTED** [review]

File: infra/gcp/terraform/k8s-infra-prow-build/prow-build/resources/monitoring/grafana/datasources.yaml (hunk 2 above)
Comment (koksay) [reply infra/gcp/terraform/k8s-infra-prow-build/prow-build/resources/monitoring/grafana/datasources.yaml:19]: right, I also did not include Loki as the logs are already shipping to GCP logging. Removing...

File: infra/gcp/terraform/k8s-infra-prow-build/prow-build/resources/monitoring/grafana/service.yaml
Comment (koksay) [reply infra/gcp/terraform/k8s-infra-prow-build/prow-build/resources/monitoring/grafana/service.yaml:1]: I suppose we don't need this one, when we create `lb.yaml`, the service will be there ClusterIP that can be used. Removing this one as well.

**Comment by upodroid** [comment]
Can we look into using Thanos to aggregate and store metrics for a period of time?

**Comment by xmudrii** [comment]
We plan to do both, however:

- This PR replicates the EKS monitoring setup pretty much 1:1 and given that we're sure that this setup works well, we don't want to change it at this point
- Integrating with Thanos will take some time and we need the monitoring stack ASAP-ish to prepare the workshop that we have at the Contributor Summit
- The way we apply resources to the GKE cluster is not really compatible with Helm (e.g. I don't know if we even have Helm in the image that's used by `post-k8sio-deploy-prow-build-resources` job), so using Helm would require additional effort that's probably not worth it as we'll have ArgoCD pipeline for the GKE cluster at some point soon-ish

That said, I have preference to proceed with this PR as it is and iterating (for both GKE and EKS clusters) in the near future.

**Review by upodroid — APPROVED** [review]
Let's add Thanos after Kubecon
/hold cancel

**Comment by BenTheElder** [comment]
Pretty sure this change is what is causing scheduling issues for some PR/release blocking jobs:

We have to be careful about increasing per-node resource utilization in the build clusters (as opposed to say, a deployment, which we can scale around just fine).
We have jobs that are best run as the sole tenant and request ~100% schedulable CPU to ensure this. See more in the linked issue.

I think we should consider reverting and then revisit shortly after, given kubernetes/kubernetes has code freeze pending and we're having trouble testing and merging PRs. https://github.com/kubernetes/test-infra/issues/32157#issuecomment-1978089591

For revisit we can discuss in https://github.com/kubernetes/test-infra/issues/32157 after confirming that CI is running smoothly again ...

**Comment by koksay** [comment]
@BenTheElder What if we remove the `requests` but keep `limits` for the node-exporter ds?

**Pull request merged**